### ISSUE TITLE
Early Stopping, Learning rate and noise decay

### ISSUE THREAD

**Issue by kim-mskw** [assume-framework/assume#398]
Implement the best practices from multi-agent Rl community and stablebaselines3 into our algorithm. Further analyse similarities between petting zoo multi-agent implementation to current RL implementation of Assume. (https://towardsdatascience.com/multi-agent-deep-reinforcement-learning-in-15-lines-of-code-using-pettingzoo-e0b963c0820b)

**Comment by mthede** [comment]
# Quick Review in Code Bases and Literature
## Early Stopping
### 1. [Stable-Baselines3](https://github.com/DLR-RM/stable-baselines3)
Three possible callbacks:
- [StopTrainingOnRewardThreshold](https://stable-baselines3.readthedocs.io/en/master/guide/callbacks.html#stoptrainingonrewardthreshold)
- [StopTrainingOnMaxEpisodes](https://stable-baselines3.readthedocs.io/en/master/guide/callbacks.html#stoptrainingonmaxepisodes)
- [StopTrainingOnNoModelImprovement](https://stable-baselines3.readthedocs.io/en/master/guide/callbacks.html#stoptrainingonnomodelimprovement)

### 2. [Standard Protocol for **Cooperative** MARL](https://arxiv.org/abs/2209.10485)
Meta-analysis on evaluation methodologies of **cooperative** MARL with proposed recommendations for standardised performance evaluation protocol. Published at NeurIPS 2022. Summary of protocol [here](https://sites.google.com/view/marl-standard-protocol#h.f998v1t8dvtq).

--> **Fixed** number of training timesteps and episodes. 

### 3. [BenchMARL (Meta Research)](https://github.com/facebookresearch/BenchMARL)
A proposed framework to deal with the fragmented community standards and reproducibility issues highlighted by the analysis above. Also some competitive environments. Published in Journal of Machine Learning Research 2024.  
> BenchMARL is a Multi-Agent Reinforcement Learning (MARL) training library created to enable reproducibility and benchmarking across different MARL algorithms and environments. Its mission is to present a standardized interface that allows easy integration of new algorithms and environments to provide a fair comparison with existing solutions. 

--> Not implemented, callbacks may be customized.

### 4. [PettingZoo](https://github.com/Farama-Foundation/PettingZoo)
--> Not implemented. 

### Summary
For comparability and according to community standards **no early stopping (as default)**
`early_stopping_steps = training_episodes / validation_episodes_interval` 
`early_stopping_threshold = 0`
 
Our current implementation of early stopping is closest to the "no model improvement" callback from SB3. However, suitable default values for steps and threshold are unclear. Should be chosen rather conservatively due to instability of environment and training, but **can be useful for experimentation or** when dealing with **time/computational restrictions**. So we'd like to **keep the options**.

_Future development_: could be restructured and generalized as callbacks.

**Comment by mthede** [comment]
## Learning Rate Decay
_Sidenote_: Currently Adam optimizer is used - Adam adapts/decays the individual learning rates of parameters automatically. However, an additional scheduling of the learning rate may still improve performance as discussed [here](https://stackoverflow.com/questions/39517431/should-we-do-learning-rate-decay-for-adam-optimizer).

### 1.1 [Stable-Baselines3](https://stable-baselines3.readthedocs.io/en/master/guide/examples.html#learning-rate-schedule)
- [General schedule function](https://github.com/DLR-RM/stable-baselines3/blob/512eea923afad6f6da4bb53d72b6ea4c6d856e59/stable_baselines3/common/utils.py#L68) that can be used for learning rate decay
- Default: constant
- Implementations: TD3 --> constant; DQN --> linear decay

### 1.2. SB3 Zoo (Training Framework)
- Linear and constant schedule used
- Hyperparameter tuning for choice of scheduler, e.g. for [PPO](https://github.com/DLR-RM/rl-baselines3-zoo/blob/726e2f1d3f1a6ea58ad4ae61c02a4ba71d241e4b/rl_zoo3/hyperparams_opt.py#L38)

### 2. Ray RL Library [Scheduler](https://docs.ray.io/en/latest/rllib/package_ref/utils.html?_gl=1*1lwhmkl*_ga*MTIxOTUwMjI5NC4xNzI3MDgyMzg5*_up*MQ..#scheduler-api)
- General scheduling capabilities
- See comment on action noise decay

### 3. [PyTorch LR Scheduler](https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate)
- 15 different schedulers available
- "Learning rate scheduling should be applied after optimizer’s update" 

### 4. [Learning to Learn Learning-Rate Schedules ](https://www.amazon.science/blog/learning-to-learn-learning-rate-schedules)
Series of conference papers on learning learning rates. Latest publication on a [GreedyLR scheduler](https://doi.org/10.1109/PRML59573.2023.10348370) that ... 

> ... outperforms several state-of-the-art schedulers in terms of accuracy, speed, and convergence. Furthermore, our method is easy to implement, computationally efficient, and requires minimal hyperparameter tuning.

This seems promising but unfortunately no code is provided. It is based on PyTorch's [ReduceLROnPlateau scheduler](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.ReduceLROnPlateau.html).

## Summary
Learning rate scheduling can be beneficial - also for Adam optimizer - and reduce runtimes. Current developments show great potential, but implementation would need to derived from publication. PyTorch offers many schedulers out of the box, but only specifically for the learning rate.

The SB3 implemetation provides a general scheduler which can be used for learning rate and action noise decay.
▶️ As we'd like to implement it anyway for the action noise decay, it can be used for the learning rate as well. It can be discussed in the future, if it should be switched to PyTorch's internal LR Scheduler.
